pkg/servicemesh/federation/status/manager.go

```diff
@@ -0,0 +1,250 @@
+}
+
+var _ Manager = (*manager)(nil)
+
+func (m *manager) FederationAdded(mesh types.NamespacedName) Handler {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if handler, exists := m.handlers[mesh]; exists {
+		m.logger.Debugf("already have a handler for mesh %s", mesh)
+		return handler
+	}
+	handler := newHandler(m, mesh)
+	m.handlers[mesh] = handler
+	return handler
+}
+
+func (m *manager) FederationDeleted(mesh types.NamespacedName) {
+	func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		if _, exists := m.handlers[mesh]; !exists {
+			m.logger.Debugf("deleting unknown handler for mesh %s", mesh)
+		}
+		delete(m.handlers, mesh)
+	}()
+
+	if err := m.PushStatus(); err != nil {
+		m.logger.Errorf("unexpected error updating FederationStatus for %s", m.name.String())
+	}
+}
+
+func (m *manager) HandlerFor(mesh types.NamespacedName) Handler {
+	handler, exists := m.handlers[mesh]
+	if !exists {
+		m.logger.Debugf("handler for mesh %s does not exist", mesh.String())
+		return nil
+	}
+	return handler
+}
+
+func (m *manager) PushStatus() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	newStatus := &v1alpha1.FederationStatus{
+		Status: v1alpha1.FederationStatusStatus{Meshes: []v1alpha1.FederationStatusDetails{}},
+	}
+	for _, handler := range m.handlers {
+		newStatus.Status.Meshes = append(newStatus.Status.Meshes, *handler.currentStatus())
+	}
+	sort.Slice(newStatus.Status.Meshes,
+		func(i, j int) bool {
+			return strings.Compare(newStatus.Status.Meshes[i].Mesh, newStatus.Status.Meshes[j].Mesh) < 0
+		})
+
+	newBytes, err := json.Marshal(newStatus)
+	if err != nil {
+		return err
+	}
+	oldBytes, err := json.Marshal(&v1alpha1.FederationStatus{
+		Status: m.status,
+	})
+	if err != nil {
+		return err
+	}
+
+	m.logger.Debugf("old bytes:\n%s\n", string(oldBytes))
+	m.logger.Debugf("new bytes:\n%s\n", string(newBytes))
+
+	patch, err := strategicpatch.CreateTwoWayMergePatchUsingLookupPatchMeta(oldBytes, newBytes, federationStatusPatchMetadata)
+	if err != nil {
+		return err
+	}
+
+	m.logger.Debugf("status patch:\n%s\n", string(patch))
+
+	if len(patch) == 0 || string(patch) == "{}" {
+		// nothing to patch
+		return nil
+	}
+
+	fedStatus, err := m.cs.CoreV1alpha1().FederationStatuses(m.name.Namespace).
+		Patch(context.TODO(), m.name.Name, types.MergePatchType, patch, metav1.PatchOptions{}, "status")
+	if err == nil {
+		m.missingCRDErrorLogged = false
+		m.status = *fedStatus.Status.DeepCopy()
+		return nil
+	} else if !(errors.IsGone(err) || errors.IsNotFound(err)) {
+		return err
+	} else if meta.IsNoMatchError(err) {
+		// The FederationStatus resource is not installed.  We won't error in this case.
+		if !m.missingCRDErrorLogged {
+			m.logger.Errorf("status will not be updated: FederationStatus CRD is not installed")
+			m.missingCRDErrorLogged = true
+		}
+		return nil
+	}
+
+	// if we get here, it means the resource doesn't exist
+	pilotPod, err := m.client.CoreV1().Pods(m.name.Namespace).Get(context.TODO(), m.name.Name, metav1.GetOptions{})
```

Comment:
do we have this info in istiod cache so that we don't need to reach the api server? Maybe this is not an issue at all, I'm just remembering the issues we had with a large number of namespaces...

Reply:
This is well scoped, but yes, I think using the lister would be more efficient.

Reply:
done